README.md
# Failmail

[![Build Status](https://travis-ci.org/hut8labs/failmail.svg?branch=master)](https://travis-ci.org/hut8labs/failmail)

Failmail is an SMTP proxy server that deduplicates and summarizes the emails it
receives. It typically sits between an application and the SMTP server that
application would use to send alerts or exception emails, and prevents the
application from overwhelming its operators with a flood of email.

Its primary goals are:

1. to help the human operators of an application or system under observation to
   understand the errors and notifications that come from the system
2. to prevent an upstream SMTP server from throttling or dropping messages
   under a high volume of errors/alerts
3. to be easier to set up and use than its predecessor,
   [failnozzle](http://github.com/wingu/failnozzle), which had strict
   assumptions about the nature and format of incoming messages

Failmail is language/framework/tool-agnostic, and usually only requires a small
configuration change in your application.


## Installation

To download and build from source:

    $ export GOPATH=...
    $ go get -u github.com/hut8labs/failmail
    $ $GOPATH/bin/failmail

Or, for 64-bit Linux, you can grab a binary of the [latest
release](https://github.com/hut8labs/failmail/releases/latest).


## Usage

See `failmail --help` for a full listing of supported options.

By default, `failmail` listens on the local port 2525
(`--bind="localhost:2525"`), and relays mail to another SMTP server (e.g.
Postfix) running on the local default SMTP port (`--relay="localhost:25"`). It
receives messages and rolls them into summaries based on their subjects,
sending a summary email out 30 seconds (`--wait=30s`) after it stops receiving
messages with those subjects, delaying no more than a total of 5 minutes
(`--max-wait=5m`). Each summary is sent to the union of all of the recipients
of the messages in the summary.

Any summary emails that it can't send via the server on port 25, it writes to a
maildir (`--fail-dir="failed"`; readable by e.g. `mutt`, or any text editor).
If the `--all-dir` option is given, `failmail` will write any email it gets to
a maildir for inspection, debugging, or archival.

### Configuration options

The value of a `failmail` setting is determined as follows:

1. If a config flag is given on the command line for that option, the value of
   that flag is used. Flags may be given with one hyphen or two (`-flag` or
   `--flag`) and with an equals sign before the value or not (`-flag=value` or
   `-flag value`).
2. Otherwise, if a config file is given (via the `--config` flag on the command
   line), and the setting is present in the file, its value from the file is
   used.
3. Finally, the `failmail` default is used. (The values of these defaults are
   shown when running `failmail --help`.)

In general, the command-line flag spells the setting name with hyphens, and the
config file uses underscores. For instance, `--max-wait=5m` is specified in a
config file as `max_wait = 5m`.

The settings are described below:

* `--all-dir` (default: none)

    write all sends to this maildir

* `--batch-expr` (default: `"{{.Header.Get \"X-Failmail-Split\"}}"`)

    an expression used to determine how messages are batched into summary emails

    (See "Configuring message batching" below.)

* `--bind-addr` (default: `"localhost:2525"`)

    local bind address

* `--bind-http` (default: `"localhost:8025"`)

    local bind address for the HTTP server

* `--config` (default: none)

    path to a config file

* `--credentials` (default: none)

    username:password for authenticating to failmail

* `--fail-dir` (default: `"failed"`)

    write failed sends to this maildir

* `--from` (default: `"failmail@$(hostname)"`)

    from address

* `--group-expr` (default: `"{{.Header.Get \"Subject\"}}"`)

    an expression used to determine how messages are grouped within summary emails

    (See "Configuring message batching" below.)

* `--max-wait` (default: `5m0s`)

    wait at most this long from first message to send summary

* `--pidfile` (default: none)

    write a pidfile to this path

* `--rate-check` (default: `1m0s`)

    how often to check whether rate limit was exceeded

* `--rate-limit` (default: `0`)

    alert if this many emails are received within a window

* `--rate-window` (default: `5`)

    the size of the rate limit window, in check intervals

* `--relay-addr` (default: `"localhost:25"`)

    relay server address

* `--relay-all`

    relay all messages to the upstream server

* `--relay-command` (default: none)

    relay messages by running this command and passing the message to stdin

* `--relay-password` (default: none)

    password for auth to relay server

* `--relay-user` (default: none)

    username for auth to relay server

* `--script` (default: none)

    SMTP session script to run

* `--shutdown-timeout` (default: `5s`)

    wait this long for open connections to finish when shutting down or reloading

* `--socket-fd` (default: `0`)

    file descriptor of socket to listen on

* `--tls-cert` (default: none)

    PEM certificate file for TLS

* `--tls-key` (default: none)

    PEM key file for TLS

* `--version`

    show the version number and exit

* `--wait-period` (default: `30s`)

    wait this long for more batchable messages

* `--write-config` (default: none)

    path to output a config file


### Configuring message batching

The `--batch-expr` and `--group-expr` options are template strings used to
determine the way that messages are batched into summary emails and grouped
within those summary emails. They are [Go
templates](http://golang.org/pkg/text/template/) evaluated in the context of a
Go [`mail.Message`](http://golang.org/pkg/net/mail/#Message). Two messages
whose expressions evaluate to the same string are treated as belonging to the
same group or batch.

Two functions are available in addition to the usual template functions:

* `match`, which takes a regular expression and a string and returns the
  leftmost match of the pattern, e.g:

          {{match "\[\w+\]" (.Header.Get "Subject")}}

    will batch messages like "[bos] error in db" and "[bos] error in web"
    together, but "[nyc] error in db" separately.

* `replace`, which takes a regular expression, a string, and a replacement, and
  returns the string with matches of the regular expression replaced by the
  replacement string, e.g.:

          {{replace "\[\w+\]" (.Header.Get "Subject") "*"}}

    will batch messages like "[bos] error in db" and "[nyc] error in db"
    together (treating them both as "[***] error in db"), but "[bos] error in
    web" separately.


## Configuration examples

See the `examples` directory for code snippets for your favorite programming
language/logging framework. (If there's one that you'd like to see, feel free
to open an issue or submit a pull request.)


## Deploying

For now, the best way to run `failmail` in production is via a tool like
`supervisord`, `runit`, or `daemontools`. Here's an example `supervisord`
configuration:

    [program:failmail]
    command=/usr/local/bin/failmail --relay="smtp.mycompany.example.com:25"
    autorestart=true
    autostart=true
    stderr_logfile=/var/log/failmail.err
    stdout_logfile=/var/log/failmail.out


## Development

    $ export GOPATH=...
    $ go get -d github.com/hut8labs/failmail
    $ cd $GOPATH/src/github.com/hut8labs/failmail
    $ ...
    $ go fmt
    $ go build
    $ go test
    $ git commit && git push ...


Other helpful testing tools include the `smtp-source` and `smtp-sink` commands,
which are part of Postfix (and are in the Debian `postfix` package):

    $ smtp-source -v -m 50 127.0.0.1:2525  # send 50 messages to failmail

The Python debugging SMTP server is also useful as an upstream server:

    python -m smtpd -n -c DebuggingServer localhost:3025


CHANGELOG.md
# 0.2.0 (2014-09-08)

## Core functionality

- Changed the way summary emails are sent to recipients. Previously, the
  summary was sent to the union of all recipients in messages that were a part of
  the summary. Now, separate summaries are sent to recipients so that they only
  see the messages that they would normally have been sent.

- Added initial support for AUTH, to require authentication from clients
  connecting to `failmail`.

- Added initial support for STARTTLS, to allow the use of encrypted connections
  to `failmail` from clients.

- Added the ability to customize the way messages are batched into summary
  messages, and grouped within summary messages, using template strings.

## For operators

- Added initial support for zero-downtime reloads/upgrades. On receiving
  SIGUSR1, `failmail` shuts itself down and hands its listening socket to a new
  child `failmail` process (which can be from a newer version) and cleanly
  shuts down, letting the new process handle all incoming connections without a
  break in service.

- Added options for config files: `--config` specifies a path to a config
  file (lines of `name = value` pairs), and `--write-config` dumps the current
  set of command line flags into a config file for editing/later use.

- Added handlers for SIGINT and SIGTERM so that the server summarizes and sends
  messages to the upstream server before terminating.

- Added a `--pidfile` flag, to write a pidfile on startup and remove it on
  shutdown.

- Added a `--version` flag, to report the version number and exit.

## For contributors

- Added more automated tests, and did some work toward end-to-end replay
  testing.

- Added the `configure` library for reading config files and creating command
  line flags dynamically from a struct definition, and binding configuration from
  a file/flags to a struct.

- Added to the `parse` library to support config file parsing.


# 0.1.0 (2014-07-20)

Initial release.


config.go
package main

import (
	"crypto/tls"
	"fmt"
	"os"
	"strings"
	"text/template"
	"time"
)

type Config struct {
	BindAddr   string        `help:"local bind address"`
	RelayAddr  string        `help:"relay server address"`
	SocketFd   int           `help:"file descriptor of socket to listen on"`
	WaitPeriod time.Duration `help:"wait this long for more batchable messages"`
	MaxWait    time.Duration `help:"wait at most this long from first message to send summary"`
	From       string        `help:"from address"`
	FailDir    string        `help:"write failed sends to this maildir"`
	AllDir     string        `help:"write all sends to this maildir"`
	RateLimit  int           `help:"alert if this many emails are received within a window"`
	RateCheck  time.Duration `help:"how often to check whether rate limit was exceeded"`
	RateWindow int           `help:"the size of the rate limit window, in check intervals"`
	BatchExpr  string        `help:"an expression used to determine how messages are batched into summary emails"`
	GroupExpr  string        `help:"an expression used to determine how messages are grouped within summary emails"`
	BindHTTP   string        `help:"local bind address for the HTTP server"`
	RelayAll   bool          `help:"relay all messages to the upstream server"`
	Pidfile    string        `help:"write a pidfile to this path"`
	Template   string        `help:"path to a summary message template file"`

	ShutdownTimeout time.Duration `help:"wait this long for open connections to finish when shutting down or reloading"`

	RelayUser     string `help:"username for auth to relay server"`
	RelayPassword string `help:"password for auth to relay server"`
	Credentials   string `help:"username:password for authenticating to failmail"`
	TlsCert       string `help:"PEM certificate file for TLS"`
	TlsKey        string `help:"PEM key file for TLS"`

	RelayCommand string `help:"relay messages by running this command and passing the message to stdin"`

	Script  string `help:"SMTP session script to run"`
	Version bool   `help:"show the version number and exit"`
}

func Defaults() *Config {
	return &Config{
		BindAddr:        "localhost:2525",
		RelayAddr:       "localhost:25",
		WaitPeriod:      30 * time.Second,
		MaxWait:         5 * time.Minute,
		From:            DefaultFromAddress("failmail"),
		FailDir:         "failed",
		RateCheck:       1 * time.Minute,
		RateWindow:      5,
		BatchExpr:       `{{.Header.Get "X-Failmail-Split"}}`,
		GroupExpr:       `{{.Header.Get "Subject"}}`,
		BindHTTP:        "localhost:8025",
		ShutdownTimeout: 5 * time.Second,
	}
}

func (c *Config) Auth() (Auth, error) {
	if c.Credentials == "" {
		return nil, nil
	}

	parts := strings.SplitN(c.Credentials, ":", 2)
	if len(parts) != 2 {
		return nil, fmt.Errorf("credentials must be in username:password format")
	}

	return &SingleUserPlainAuth{Username: parts[0], Password: parts[1]}, nil
}

func (c *Config) Batch() GroupBy {
	return GroupByExpr("batch", c.BatchExpr)
}

func (c *Config) Group() GroupBy {
	return GroupByExpr("group", c.GroupExpr)
}

func (c *Config) Upstream() (Upstream, error) {
	var upstream Upstream
	if c.RelayAddr == "debug" {
		upstream = &DebugUpstream{os.Stdout}
	} else {
		upstream = &LiveUpstream{logger("upstream"), c.RelayAddr, c.RelayUser, c.RelayPassword}
	}

	if c.AllDir != "" {
		allMaildir := &Maildir{Path: c.AllDir}
		if err := allMaildir.Create(); err != nil {
			return upstream, err
		}
		upstream = NewMultiUpstream(&MaildirUpstream{allMaildir}, upstream)
	}

	if c.RelayCommand != "" {
		upstream = NewMultiUpstream(&ExecUpstream{c.RelayCommand}, upstream)
	}
	return upstream, nil
}

func (c *Config) TLSConfig() (*tls.Config, error) {
	if c.TlsCert == "" || c.TlsKey == "" {
		return nil, nil
	}

	cert, err := tls.LoadX509KeyPair(c.TlsCert, c.TlsKey)
	if err != nil {
		return nil, err
	}
	return &tls.Config{Certificates: []tls.Certificate{cert}}, nil
}

func (c *Config) Socket() (ServerSocket, error) {
	if c.SocketFd > 0 {
		return NewFileServerSocket(uintptr(c.SocketFd))
	}
	return NewTCPServerSocket(c.BindAddr)
}

func (c *Config) SummaryRenderer() SummaryRenderer {
	if c.Template != "" {
		tmpl := template.Must(template.New(c.Template).Funcs(SUMMARY_TEMPLATE_FUNCS).ParseFiles(c.Template))
		return &TemplateRenderer{tmpl}
	}
	return &DefaultRenderer{}
}
